### PR TITLE
Add `FileSegmentRef`

### PR DESCRIPTION
**Pull request by tzerrell**
Add a variety of `SegmentRef` that saves the segment to a file (at a user-specified location). Also, use this `FileSegmentRef` in the EVM example.

### CODE REVIEW

File: examples/evm/src/main.rs [hunk 1]
```diff
@@ -86,7 +90,7 @@ async fn main() {
         .add_input(&to_vec(&zkdb).unwrap())
         .build();
     let mut exec = Executor::from_elf(env, EVM_ELF).unwrap();
-    let session = exec.run().unwrap();
+    let session = exec.run_with_callback(|segment| Ok(Box::new(FileSegmentRef::new(&segment, &temp_dir())?))).unwrap();
```
Comment (mothran) [inline examples/evm/src/main.rs:93]: I think this calls temp_dir() each callback so each segment would get its own new directory. I would consider calling it once before run_with_callback and then just passing a reference on each callback.

File: risc0/zkvm/src/session.rs [hunk 2]
```diff
@@ -155,3 +160,36 @@ impl SimpleSegmentRef {
...
+        file.read_to_end(&mut contents)?;
+        let segment: Segment = bincode::deserialize(&contents)?;
+        Ok(segment)
+    }
+}
+
+impl FileSegmentRef {
+    /// Construct a [FileSegmentRef]
+    ///
+    /// This builds a FileSegmentRef that stores `segment` in a file at `path`.
```
Comment (mothran) [inline risc0/zkvm/src/session.rs:187]: This is a little bit of a odd suggestion but should we consider linking to the EVM code in the docstr kinda like: "if you wanna see an example of using this check here ...." ?

**Review by mothran — CHANGES_REQUESTED** [review]
Thank you! Only minor tweaks in the comments from me.

File: risc0/zkvm/src/session.rs (hunk 2 above)
Comment (tzerrell) [reply risc0/zkvm/src/session.rs:187]: Interesting! Perhaps a bit odd, but probably a discoverability win so I like it :+1:

**Review by mothran — APPROVED** [review]
Looks good! Does it have the memory overhead properties you were hoping for? I can run `heaptrack` on it if you need.

**Comment by tzerrell** [comment]
> Looks good! Does it have the memory overhead properties you were hoping for? I can run `heaptrack` on it if you need.

I'm still testing this version, but an earlier version ([da2b55d](https://github.com/risc0/risc0/pull/560/commits/da2b55d88dac58b0716f6a56d22aff5fc98844a9)) was an approximately constant ~12GB for runs between 1.5M and 56M cycles, IIRC. So yes, this seems to be working as intended in that regard.

File: examples/evm/src/main.rs [hunk 3]
```diff
@@ -86,7 +86,10 @@ async fn main() {
         .add_input(&to_vec(&zkdb).unwrap())
         .build();
     let mut exec = Executor::from_elf(env, EVM_ELF).unwrap();
-    let session = exec.run().unwrap();
+    let segment_dir = temp_dir();
```
Comment (flaub) [inline examples/evm/src/main.rs:89]: I think we should use the crate `tempfile` so that the tempdir gets removed when it goes out of scope.

**Pull request merged**